File: Python/49. Group Anagrams.py
from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = defaultdict(list) #map the charcount against list of anagrams
        for s in strs:
            charcount = [0]*26  #list with a-z
            for c in s:
                charcount[ord(c)-ord('a')] += 1
            res[tuple(charcount)].append(s)
        print(res)
        return res.values()
    

print(Solution().groupAnagrams(["eat","tea","tan","ate","nat","bat"]))

File: Typescript/49. Group Anagrams.ts
function groupAnagrams(strs: string[]): string[][] {
  const res: Map<string, string[]> = new Map();
  for (const s of strs) {
    const charcount: number[] = new Array(26).fill(0);
    for (const c of s) {
      charcount[c.charCodeAt(0) - "a".charCodeAt(0)]++;
    }
    const key: string = charcount.join();
    if (!res.has(key)) {
      res.set(key, []);
    }
    res.get(key)!.push(s);
  }
  return Array.from(res.values());
}

console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]));
console.log(groupAnagrams([""]));
console.log(groupAnagrams(["a"]));
